mmengine/utils/progressbar_rich.py

```diff
@@ -0,0 +1,312 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import multiprocessing
+import time
+from collections.abc import Iterable
+from functools import partial
+from multiprocessing import Pool
+from typing import List
+
+import rich
+from rich.progress import Progress
+
+from .timer import Timer
+
+
+class RichProgressBar:
+    """use rich to enhance progress bar."""
+
+    def __init__(self):
+        self.bar = Progress()
+        self.tasks = []
+        self.descriptions = []
+        self.colors = []
+        self.bar.start()
+
+    def __del__(self):
+        self.bar.stop()
+
+    def write(self, msg, color='blue'):
+        rich.print(f'[{color}]{msg}')
+
+    def add_multi_tasks(self, total, color='blue', description='Process...'):
+        if total <= 0:
+            raise ValueError('total should be greater than 0')
+        self.descriptions.append(description)
+        self.colors.append(color)
+        self.tasks.append(
+            self.bar.add_task(
+                f'[{color}]{description}_0/{total}', total=total))
+
+    def add_single_task(self,
+                        total=None,
+                        color='blue',
+                        description='Process...'):
+        assert len(self.tasks) == 0
+        self.descriptions.append(description)
+        self.colors.append(color)
+        if total is not None:
+            if total <= 0:
+                raise ValueError(
+                    'Total only exists if it is greater than zero or None.')
+            self.tasks.append(
+                self.bar.add_task(
+                    f'[{color}]{description}_0/{total}', total=total))
+            self.infinite = False
+        else:
+            self.write('completed: 0, elapsed: 0s', self.colors[0])
+            self.timer = Timer()
+            self.infinite = True
+            self.completed = 0
+            self.tasks.append(0)
```

Comment:
BTW, suggest initializing `self.timer=None` and `self.completed` in `__init__`

```suggestion
def add_task(self,
             total=None,
             color='blue',
             description='Process...')
    if total is not None:
        assert not self.infinite, ('The prior task is an infinite task (total is None), RichProgressBar can only accept one infinite task')
        if total <= 0:
            raise ValueError(
                'Total only exists if it is greater than zero or None.')
        self.tasks.append(
            self.bar.add_task(
                f'[{color}]{description}_0/{total}', total=total))
    else:
        assert not self.tasks, ('Since the total argument is None, this task is considered infinite. RichProgressBar should not have any other tasks added to it when an infinite task is present.')
        self.write('completed: 0, elapsed: 0s', self.colors[0])
        self.timer = Timer()
        self.infinite = True
        self.completed = 0
        self.tasks.append(0)
```